File: BOJ/Binary Search/6236-용돈관리.java
import java.util.*;
import java.lang.*;
import java.io.*;

class Main {
    static int n, m;
    static int[] arr;

    public static void main(String[] args) throws IOException{
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        n = Integer.parseInt(st.nextToken());
        m = Integer.parseInt(st.nextToken());

        arr = new int[n];
        int max = 0; // max 가 최소 k
        int total = 0; // total 값이 최대 k

        for (int i = 0; i < n; i++) {
            st = new StringTokenizer(br.readLine());
            arr[i] = Integer.parseInt(st.nextToken());
            max = Math.max(max, arr[i]);
            total += arr[i];
        }

        int start = max;
        int end = total;
        int r = total;

        while (start <= end) {
            int mid = (start + end) / 2;
            int cnt = 1;
            int money = mid; // 인출 금액

            for (int i = 0; i < n; i++) {
                if (money < arr[i]) {
                    cnt++;
                    money = mid;
                }
                money -= arr[i];
            }

            if (cnt <= m) {
                r = mid;
                end = mid - 1;
            } else {
                start = mid + 1;
            }
        }

        System.out.println(r);
    }
}

File: BOJ/Binary Search/6236-용돈관리.py
import sys

input = sys.stdin.readline

n, m = map(int, input().split())
arr = []
for _ in range(n):
    arr.append(int(input()))

start = max(arr)
end = sum(arr)
answer = end

while start <= end:
    mid = (start + end) // 2

    cnt = 1
    cur = mid
    for num in arr:
        if cur < num:
            cnt += 1
            cur = mid
        cur -= num
        if cnt > m:
            break
    if cnt <= m:
        answer = mid
        end = mid - 1
    else:
        start = mid + 1

print(answer)